gpt.lua

```diff
@@ -0,0 +1,555 @@
+    retries = 0
+    if current_status == Status.waiting then
+      set_current_status(Status.receiving)
+    elseif current_status ~= Status.receiving then
+      qerror('Error: data received by polling while status was ' .. string_from_Status(current_status))
+    end
+
+    local sanitized_data = sanitize_response(data)
+
+    if string.find(data, "Excerpt") then
+      total_data = total_data .. NEWLINE .. NEWLINE .. sanitized_data
+    else
+      total_data = total_data .. NEWLINE .. sanitized_data
+    end
+
+    gui_text = total_data
+  else
+    if current_status == Status.receiving then
+      if retries >= max_retries then
+        debug_log("Max retries reached.")
+        retries = 0
+        stop_polling(client)
+        return
+      else
+        retries = retries + 1
+      end
+    elseif current_status == Status.waiting then
+      update_progress_indicator()
+    end
+  end
+
+  if os.difftime(os.time(), start_time) >= timeout then
+    debug_log('Reached time limit of ' .. timeout .. ', stopping polling.')
+    retries = 0
+    stop_polling(client)
+    return
+  end
+end
+
+-- Sends json request to the remote service helper.
+local function send(request)
+  set_current_status(Status.waiting)
+  client = make_client()
+  start_time = os.time()
+  debug_log('Sending request... \n')
+  client:send(request)
+  poll(client)
+end
+
+-- Primary entrypoint to the script's functionality. Initiates a check
+-- of the UI to see if a supported written content item is being displayed.
+-- If so, then submit a request to the remote helper script.
+function fetch_generated_text()
+  skip = skip + 1
+  if skip < 20 then return end
+  skip = 0
+
+  if current_status ~= Status.start then
+    debug_log("Current status was not start status, aborting. Status was: " .. string_from_Status(current_status))
+    return
+  end
+
+  local knowledge_description, content_type = knowledge_description()
+
+  if knowledge_description == last_knowledge_description then
+    return
+  end
+
+  if not knowledge_description then
+    debug_log('Poem description became nil, retrying...')
+    last_knowledge_description = nil
+    return
+  end
+
+  if content_type == Content_Type.unsupported then
+    gui_text = "This content type is not supported. Please select a " .. valid_content_type_list .. "."
+    last_knowledge_description = nil
+    return
+  end
+
+  debug_log('Got new ' .. content_type .. " description: " .. knowledge_description .. "\n")
+  last_knowledge_description = knowledge_description
+  gui_text = "Generating text from description, please wait...  "
+  debug_log("Submitting request to OpenAI remote service... \n")
+  send(request_from(knowledge_description, content_type))
+end
+
+--
+-- GUI: Overlay
+--
+
+GPTBannerOverlay = defclass(GPTBannerOverlay, overlay.OverlayWidget)
+GPTBannerOverlay.ATTRS{
+    default_pos={x=-35,y=-2},
+    default_enabled=true,
+    viewscreens={'dwarfmode/ViewSheets/UNIT','dwarfmode/ViewSheets/ITEM'},
```

Comment:
this seems too broad -- do we need more fine-grained focus strings here? it's not hard to add more context in `modules/Gui.cpp`

Reply:
Just using what was available. If you give me one for the UNIT/Skills/Knowledge I'll gladly use it

Reply:
yeah, let me see if I can give you something better here.

Reply:
here we go:
![image](https://github.com/DFHack/scripts/assets/977482/f86ef1f2-58cb-4b7c-b5da-a8a43120f071)
and with a knowledge item selected:
![image](https://github.com/DFHack/scripts/assets/977482/eb78fc4a-6365-42d1-9400-5c4fa2d5c58e)